### ISSUE TITLE
[BUG] VM can not migrate during upgrade

### ISSUE THREAD

**Issue by Vicente-Cheng** [harvester/harvester#5076]
**Describe the bug**
VM can not migrate during upgrade.
for example, I run the VM on node1. when node1 is draining, it should migrate to other nodes but fail.
That will impact the service, which should not be down.

**To Reproduce**
1. create 3 nodes cluster (version v1.2.1)
2. create guest cluster with rke2 cloud provider
3. upgrade to v1.3.0-rc1
4. Check the VM should running after upgrade.

**Expected behavior**
VM should migrate successfully

**Support bundle**
None

**Environment**
 - Harvester ISO version: v1.2.1 -> v1.3.0-rc1
 - Underlying Infrastructure (e.g. Baremetal with Dell PowerEdge R630): kvm, 3 nodes cluster

**Additional context**
Some migrate failed logs from virt-handler.
```
2024-01-31T13:23:48.555760707Z stderr F {"component":"virt-handler","level":"info","msg":"re-enqueuing VirtualMachineInstance default/rke2-guest-01-pool1-b4649f31-9g2nx","pos":"vm.go:1588","reason":"syncing migration target failed: unknown error encountered sending command SyncMigrationTarget: rpc error: code = Unavailable desc = error reading from server: EOF","timestamp":"2024-01-31T13:23:48.555655Z"}
...
2024-01-31T13:25:33.067561695Z stderr F {"component":"virt-handler","level":"info","msg":"re-enqueuing VirtualMachineInstance default/rke2-guest-01-pool1-b4649f31-9g2nx","pos":"vm.go:1588","reason":"syncing migration target failed: unknown error encountered sending command SyncMigrationTarget: rpc error: code = Unavailable desc = error reading from server: EOF","timestamp":"2024-01-31T13:25:33.067501Z"}
...
2024-01-31T13:34:22.398539648Z stderr F {"component":"virt-handler","level":"info
","msg":"re-enqueuing VirtualMachineInstance default/rke2-guest-01-pool1-b4649f31-9g2nx","pos":"vm.go:1588","reason":"syncing migration target failed: unknown error encountered sending command SyncMigrationTarget: rpc error: code = Unavailable desc = error reading from server: EOF","timestamp":"2024-01-31T13:34:22.398483Z"}
...
2024-01-31T13:40:03.780008282Z stderr F {"component":"virt-handler","level":"info","msg":"re-enqueuing VirtualMachineInstance default/rke2-guest-01-pool1-b4649f31-9g2nx","pos":"vm.go:1588","reason":"Can not update a VirtualMachineInstance with unresponsive command server.","timestamp":"2024-01-31T13:40:03.779846Z"}
...
2024-01-31T13:42:05.481328521Z stderr F {"component":"virt-handler","kind":"","level":"info","msg":"Performing final local cleanup for vmi with uid d8cb8184-dc73-4f2f-b9b0-c22d204d9757","name":"rke2-guest-01-pool1-b4649f31-9g2nx","namespace":"default","pos":"vm.go:2062","timestamp":"2024-01-31T13:42:05.480996Z","uid":"d8cb8184-dc73-4f2f-b9b0-c22d204d9757"}
```

**Comment by bk201** [comment]
I can also reproduce this when upgrading from v1.2.1 to v1.2-head (with the [PR](https://github.com/harvester/harvester-installer/pull/646)).

**Comment by bk201** [comment]
Some observation:
- VMs with a management network only can migrate without issue.
- VMs with VLAN network can't migrate, the destination pod shows the error:

<details>
<summary>click to expand</summary>

```
compute {"component":"virt-launcher-monitor","level":"error","msg":"could not create up serial console term file: /var/run/kubevirt-private/065ef16b-268f-4e5d-814a-f100076373d6/virt-serial0
-log-sigTerm","pos":"virt-launcher-monitor.go:72","reason":"open /var/run/kubevirt-private/065ef16b-268f-4e5d-814a-f100076373d6/virt-serial0-log-sigTerm: no such file or directory","timesta
mp":"2024-02-04T15:31:40.290611Z"}
compute {"component":"virt-launcher","level":"info","msg":"Collected all requested hook sidecar sockets","pos":"manager.go:86","timestamp":"2024-02-04T15:31:40.333341Z"}
compute {"component":"virt-launcher","level":"info","msg":"Sorted all collected sidecar sockets per hook point based on their priority and name: map[]","pos":"manager.go:89","timestamp":"20
24-02-04T15:31:40.333387Z"}
compute {"component":"virt-launcher","level":"info","msg":"Connecting to libvirt daemon: qemu+unix:///session?socket=/var/run/libvirt/virtqemud-sock","pos":"libvirt.go:563","timestamp":"202
4-02-04T15:31:40.333712Z"}
compute {"component":"virt-launcher","level":"info","msg":"Connecting to libvirt daemon failed: virError(Code=38, Domain=7, Message='Failed to connect socket to '/var/run/libvirt/virtqemud-
sock': No such file or directory')","pos":"libvirt.go:571","timestamp":"2024-02-04T15:31:40.334043Z"}
compute {"component":"virt-launcher","level":"info","msg":"libvirt version: 9.0.0","subcomponent":"libvirt","thread":"41","timestamp":"2024-02-04T15:31:40.446000Z"}
compute {"component":"virt-launcher","level":"info","msg":"hostname: ubuntu","subcomponent":"libvirt","thread":"41","timestamp":"2024-02-04T15:31:40.446000Z"}
compute {"component":"virt-launcher","level":"error","msg":"internal error: Unable to get session bus connection: Cannot spawn a message bus without a machine-id: Invalid machine ID in /var
/lib/dbus/machine-id or /etc/machine-id","pos":"virGDBusGetSessionBus:128","subcomponent":"libvirt","thread":"41","timestamp":"2024-02-04T15:31:40.446000Z"}
compute {"component":"virt-launcher","level":"error","msg":"internal error: Unable to get system bus connection: Could not connect: No such file or directory","pos":"virGDBusGetSystemBus:10
1","subcomponent":"libvirt","thread":"41","timestamp":"2024-02-04T15:31:40.446000Z"}
compute {"component":"virt-launcher","level":"info","msg":"Connected to libvirt daemon","pos":"libvirt.go:579","timestamp":"2024-02-04T15:31:40.837457Z"}
compute {"component":"virt-launcher","level":"info","msg":"Registered libvirt event notify callback","pos":"client.go:562","timestamp":"2024-02-04T15:31:40.848909Z"}
compute {"component":"virt-launcher","level":"info","msg":"Marked as ready","pos":"virt-launcher.go:75","timestamp":"2024-02-04T15:31:40.849312Z"}
compute {"component":"virt-launcher","kind":"","level":"info","msg":"Executing PreStartHook on VMI pod environment","name":"ubuntu-0","namespace":"harvester-public","pos":"manager.go:671","
timestamp":"2024-02-04T15:31:41.661386Z","uid":"065ef16b-268f-4e5d-814a-f100076373d6"}
compute {"component":"virt-launcher","kind":"","level":"info","msg":"Starting PreCloudInitIso hook","name":"ubuntu-0","namespace":"harvester-public","pos":"manager.go:692","timestamp":"2024
-02-04T15:31:41.661455Z","uid":"065ef16b-268f-4e5d-814a-f100076373d6"}
compute {"component":"virt-launcher","level":"error","msg":"failed to get a link for interface: ","pos":"generators.go:132","reason":"Link not found","timestamp":"2024-02-04T15:31:41.661845
Z"}
compute {"component":"virt-launcher","level":"fatal","msg":"failed to create libvirt configuration","pos":"podnic.go:95","reason":"Link not found","timestamp":"2024-02-04T15:31:41.661886Z"}
compute panic: failed to create libvirt configuration
compute {"component":"virt-launcher-monitor","level":"info","msg":"Reaped pid 12 with status 512","pos":"virt-launcher-monitor.go:192","timestamp":"2024-02-04T15:31:41.666790Z"}
compute {"component":"virt-launcher-monitor","level":"error","msg":"dirty virt-launcher shutdown: exit-code 2","pos":"virt-launcher-monitor.go:210","timestamp":"2024-02-04T15:31:41.666856Z"
}
compute {"component":"virt-launcher-monitor","level":"error","msg":"failed to open file /var/run/kubevirt/passt.log","pos":"virt-launcher-monitor.go:421","reason":"open /var/run/kubevirt/pa
sst.log: no such file or directory","timestamp":"2024-02-04T15:31:41.666881Z"}
```

</details>

**Comment by ibrokethecloud** [comment]
I noticed mutlus is not ready during the upgrade as the chart upgrade keeps failing as underlying node is unable to meet the minimum version check. I am not yet sure if this is related. 

```
vm1:/var/run/kubevirt # k logs -f helm-install-rke2-multus-42dpc -n kube-system
if [[ ${KUBERNETES_SERVICE_HOST} =~ .*:.* ]]; then
	echo "KUBERNETES_SERVICE_HOST is using IPv6"
	CHART="${CHART//%\{KUBERNETES_API\}%/[${KUBERNETES_SERVICE_HOST}]:${KUBERNETES_SERVICE_PORT}}"
else
	CHART="${CHART//%\{KUBERNETES_API\}%/${KUBERNETES_SERVICE_HOST}:${KUBERNETES_SERVICE_PORT}}"
fi

set +v -x
+ [[ true != \t\r\u\e ]]
+ [[ '' == \1 ]]
+ [[ '' == \v\2 ]]
+ [[ -f /config/ca-file.pem ]]
+ [[ -n '' ]]
+ shopt -s nullglob
+ helm_content_decode
+ set -e
+ ENC_CHART_PATH=/chart/rke2-multus.tgz.base64
+ CHART_PATH=/tmp/rke2-multus.tgz
+ [[ ! -f /chart/rke2-multus.tgz.base64 ]]
+ base64 -d /chart/rke2-multus.tgz.base64
+ CHART=/tmp/rke2-multus.tgz
+ set +e
+ [[ install != \d\e\l\e\t\e ]]
+ helm_repo_init
+ grep -q -e 'https\?://'
+ [[ helm_v3 == \h\e\l\m\_\v\3 ]]
+ [[ /tmp/rke2-multus.tgz == stable/* ]]
+ [[ -n '' ]]
+ helm_update install --set-string global.clusterCIDR=10.52.0.0/16 --set-string global.clusterCIDRv4=10.52.0.0/16 --set-string global.clusterDNS=10.53.0.10 --set-string global.clusterDomain=cluster.local --set-string global.rke2DataDir=/var/lib/rancher/rke2 --set-string global.serviceCIDR=10.53.0.0/16
+ [[ helm_v3 == \h\e\l\m\_\v\3 ]]
++ jq -r '"\(.[0].app_version),\(.[0].status)"'
++ tr '[:upper:]' '[:lower:]'
++ helm_v3 ls --all -f '^rke2-multus$' --namespace kube-system --output json
+ LINE=3.9.3,deployed
+ IFS=,
+ read -r INSTALLED_VERSION STATUS _
+ VALUES=
+ [[ install = \d\e\l\e\t\e ]]
+ [[ 3.9.3 =~ ^(|null)$ ]]
+ [[ deployed =~ ^(pending-install|pending-upgrade|pending-rollback)$ ]]
+ [[ deployed == \d\e\p\l\o\y\e\d ]]
+ echo 'Already installed rke2-multus'
Already installed rke2-multus
+ [[ helm_v3 == \h\e\l\m\_\v\3 ]]
+ helm_v3 mapkubeapis rke2-multus --namespace kube-system
2024/02/05 03:19:37 Release 'rke2-multus' will be checked for deprecated or removed Kubernetes APIs and will be updated if necessary to supported API versions.
2024/02/05 03:19:37 Get release 'rke2-multus' latest version.
2024/02/05 03:19:37 Check release 'rke2-multus' for deprecated or removed APIs...
2024/02/05 03:19:37 Finished checking release 'rke2-multus' for deprecated or removed APIs.
2024/02/05 03:19:37 Release 'rke2-multus' has no deprecated or removed APIs.
2024/02/05 03:19:37 Map of release 'rke2-multus' deprecated or removed APIs to supported versions, completed successfully.
+ echo 'Upgrading helm_v3 chart'
Upgrading rke2-multus
+ echo 'Upgrading rke2-multus'
+ shift 1
+ helm_v3 upgrade --set-string global.clusterCIDR=10.52.0.0/16 --set-string global.clusterCIDRv4=10.52.0.0/16 --set-string global.clusterDNS=10.53.0.10 --set-string global.clusterDomain=cluster.local --set-string global.rke2DataDir=/var/lib/rancher/rke2 --set-string global.serviceCIDR=10.53.0.0/16 rke2-multus /tmp/rke2-multus.tgz
Error: UPGRADE FAILED: chart requires kubeVersion: >= v1.27.9 which is incompatible with Kubernetes v1.25.9+rke2r1
+ exit
```

**Comment by Vicente-Cheng** [comment]
Confirm that we can reproduce it only bump kubevirt to v1.1.0

**Comment by ibrokethecloud** [comment]
The error is being generated during migration as lookup of nic fails in the old pod.

The error in the new launcher pod:

```
{"component":"virt-launcher","level":"error","msg":"failed to get a link for interface: ","pos":"generators.go:132","reason":"Link not found","timestamp":"2024-02-05T22:47:32.399415Z"}
{"component":"virt-launcher","level":"fatal","msg":"failed to create libvirt configuration","pos":"podnic.go:95","reason":"Link not found","timestamp":"2024-02-05T22:47:32.399531Z"}
```

is being returned from here: https://github.com/kubevirt/kubevirt/blob/v1.1.0/pkg/network/setup/podnic.go#L95

As evident there is no interface name specified in the error and this is because of the logic used to identify the interface name:

https://github.com/kubevirt/kubevirt/blob/v1.1.0/pkg/network/setup/podnic.go#L54

The `link.DiscoverByNetwork` seems to eat up the link not found error earlier: https://github.com/kubevirt/kubevirt/blob/v1.1.0/pkg/network/setup/podnic.go#L60

error being ignored: https://github.com/kubevirt/kubevirt/blob/main/pkg/network/link/discovery.go#L66

As a result the podNIC handler is setup with `podnic.podInterfaceName = ""`

```
if ifaceLink == nil {
		podnic.podInterfaceName = ""
	} else {
		podnic.podInterfaceName = ifaceLink.Attrs().Name
	}
```


When the bridge generator is called, it fails:

https://github.com/kubevirt/kubevirt/blob/v1.1.0/pkg/network/domainspec/generators.go#L130

as a result VM migration fails.

A possible work around may be to set the mutlus network as the default network to default:

https://github.com/kubevirt/api/blob/main/core/v1/schema.go#L1549

This will cause the method `HashedPodInterfaceName` to return default interface eth0, which would be available in older generation of launcher pods 
https://github.com/kubevirt/kubevirt/blob/main/pkg/network/namescheme/networknamescheme.go#L58 

However this will mean users need to edit all their pods pre-upgrade, and still needs an outage.

As a result it might be better to just advise users to shutdown all workloads pre-upgrade to allow a seamless upgrade, and post upgrade the new interface naming convention will kick in the launcher pods.

**Comment by lanfon72** [comment]
`not-require/test-plan` added, we discussed this issue in https://github.com/harvester/tests/issues/1097 and would use it to track the implementation.

**Comment by Vicente-Cheng** [comment]
related to issue: https://github.com/kubevirt/kubevirt/issues/11677

**Comment by Vicente-Cheng** [comment]
Depends on https://github.com/harvester/harvester/pull/5166
After the above PR merged, we could backport it to v1.2-head

**Comment by lanfon72** [comment]
Verified this bug has been fixed.

Test Information
----
* Environment: **baremetal DL360G9 3-4 nodes**
* Harvester Version: **v1.2.1** -> **v1.2.2-rc1** -> **v1.3-d2f04e5a-head**
* **ui-source** Option: **Auto**

### Verify Steps
1. Install Harvester with any nodes
2. Create an image for VM creation
3. Create cluster network and VM network
4. Create VM `vm1` with VLAN NIC
5. Perform upgrade
6. Upgrade should success
7. `vm1` should automatically migrate to other nodes during upgrade
8. `vm1` should operate normally
